<a href="https://colab.research.google.com/github/Angelalena23/Python-Projects/blob/main/Technical_Test_Python_Angelica_Magdalena.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Preparation

**DATASET**

Bank Churner Data : https://docs.google.com/spreadsheets/d/1N9NnX8YsIlY3cR4DObp2VuVx7qFr1FCHmhpswyQDK2o/edit?usp=sharing

Credit Limit : https://docs.google.com/spreadsheets/d/1IKWyQqOXHJE5lvoDMM86xPeDNjqMXzpiFS26HocAXR0/edit?usp=sharing


In [79]:
import pandas as pd                 # Library used for working with data sets and perform data analysis.
import numpy as np                  # To perform mathematical operasion and statistics
from datetime import datetime       # Used to perform dates and time manipulation

In [80]:
# Import Dataset
# Bank Churner Data
sheet_url = 'https://docs.google.com/spreadsheets/d/1Z8eV1mJZEQiNoBPnypZF1tUrrrpyxmwtPmFSnQbMweA/edit#gid=2084583872'
sheet_url_trf = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')
BCD = pd.read_csv(sheet_url_trf)
BCD.head()

Unnamed: 0  CLIENTNUM     Attrition_Flag  Customer_Age Gender  \
0           0  768805383  Existing Customer         45.00      M   
1           1  818770008  Existing Customer         49.00      F   
2           2  713982108  Existing Customer         51.00      M   
3           3  769911858  Existing Customer         40.00      F   
4           4  709106358  Existing Customer         40.00      M   

   Dependent_count Education_Level Marital_Status Income_Category  \
0                3     High School        Married     $60K - $80K   
1                5        Graduate         Single  Less than $40K   
2                3        Graduate        Married    $80K - $120K   
3                4     High School        Unknown  Less than $40K   
4                3      Uneducated        Married     $60K - $80K   

  Card_Category  Months_on_book  Total_Relationship_Count  \
0          Blue              39                         5   
1          Blue              44                         6   
2          Blue              36                         4   
3          Blue              34                         3   
4          Blue              21                         5   

   Months_Inactive_12_mon  Contacts_Count_12_mon  Credit_Limit  
0                       1                      3      12691.00  
1                       1                      2       8256.00  
2                       1                      0       3418.00  
3                       4                      1       3313.00  
4                       1                      0       4716.00

In [81]:
# delete column unnamed on Bank churn data
column_numbers = [x for x in range(BCD.shape[1])]
column_numbers .remove(0)
BCD_cleaned = BCD.iloc[:, column_numbers]

In [82]:
# Check data type
BCD_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10186 entries, 0 to 10185
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CLIENTNUM                 10186 non-null  int64  
 1   Attrition_Flag            10186 non-null  object 
 2   Customer_Age              10162 non-null  float64
 3   Gender                    10177 non-null  object 
 4   Dependent_count           10186 non-null  int64  
 5   Education_Level           10186 non-null  object 
 6   Marital_Status            10186 non-null  object 
 7   Income_Category           10186 non-null  object 
 8   Card_Category             10186 non-null  object 
 9   Months_on_book            10186 non-null  int64  
 10  Total_Relationship_Count  10186 non-null  int64  
 11  Months_Inactive_12_mon    10186 non-null  int64  
 12  Contacts_Count_12_mon     10186 non-null  int64  
 13  Credit_Limit              10186 non-null  float64
dtypes: flo

In [83]:
BCD_cleaned.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer         45.00      M                3   
1  818770008  Existing Customer         49.00      F                5   
2  713982108  Existing Customer         51.00      M                3   
3  769911858  Existing Customer         40.00      F                4   
4  709106358  Existing Customer         40.00      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  Total_Relationship_Count  Months_Inactive_12_mon  \
0              39                         5                       1   
1              44                         6                       1   
2              36                         4                       1   
3              34                         3                       4   
4              21                         5                       1   

   Contacts_Count_12_mon  Credit_Limit  
0                      3      12691.00  
1                      2       8256.00  
2                      0       3418.00  
3                      1       3313.00  
4                      0       4716.00

In [84]:
# Import Dataset
# Credit Limit
sheet_url = 'https://docs.google.com/spreadsheets/d/1LUjgxIPE6dpys8ju1QeqasB1-1CAdpKVy-Z5uCYsNxU/edit#gid=1594162646'
sheet_url_trf = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')
CL = pd.read_csv(sheet_url_trf)
CL.head()

CLIENTNUM  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0  768805383      12691.00                  777         11914.00   
1  818770008       8256.00                  864          7392.00   
2  713982108       3418.00                    0          3418.00   
3  769911858       3313.00                 2517           796.00   
4  709106358       4716.00                    0          4716.00   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                  1.33             1144              42                 1.62   
1                  1.54             1291              33                 3.71   
2                  2.59             1887              20                 2.33   
3                  1.41             1171              20                 2.33   
4                  2.17              816              28                 2.50   

   Avg_Utilization_Ratio  \
0                   0.06   
1                   0.10   
2                   0.00   
3                   0.76   
4                   0.00   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                               0.00                                                                                    
1                                               0.00                                                                                    
2                                               0.00                                                                                    
3                                               0.00                                                                                    
4                                               0.00                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                               1.00                                                                                   
1                                               1.00                                                                                   
2                                               1.00                                                                                   
3                                               1.00                                                                                   
4                                               1.00

# Join the dataset

In [85]:
#Merge data
df_merge = BCD_cleaned.merge(CL, how = 'inner', left_on=['CLIENTNUM'], right_on=['CLIENTNUM'])

df_merge.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer         45.00      M                3   
1  768805383  Existing Customer         30.00      F                3   
2  768805383  Existing Customer         45.00      M                3   
3  818770008  Existing Customer         49.00      F                5   
4  713982108  Existing Customer         51.00      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1     High School        Married   60𝐾− 80K          Blue   
2     High School         Single   60𝐾− 80K          Blue   
3        Graduate         Single  Less than $40K          Blue   
4        Graduate        Married    $80K - $120K          Blue   

   Months_on_book  ...  Credit_Limit_y  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...        12691.00                  777         11914.00   
1              39  ...        12691.00                  777         11914.00   
2              39  ...        12691.00                  777         11914.00   
3              44  ...         8256.00                  864          7392.00   
4              36  ...         3418.00                    0          3418.00   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                  1.33             1144              42                 1.62   
1                  1.33             1144              42                 1.62   
2                  1.33             1144              42                 1.62   
3                  1.54             1291              33                 3.71   
4                  2.59             1887              20                 2.33   

   Avg_Utilization_Ratio  \
0                   0.06   
1                   0.06   
2                   0.06   
3                   0.10   
4                   0.00   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                               0.00                                                                                    
1                                               0.00                                                                                    
2                                               0.00                                                                                    
3                                               0.00                                                                                    
4                                               0.00                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                               1.00                                                                                   
1                                               1.00                                                                                   
2                                               1.00                                                                                   
3                                               1.00                                                                                   
4                                               1.00                                                                                   

[5 rows x 24 columns]

In [86]:
df_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10186 entries, 0 to 10185
Data columns (total 24 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10186 non-null  int64  
 1   Attrition_Flag                                                                                                                      10186 non-null  object 
 2   Customer_Age                                                                                                                        10162 non-null  float64
 3   Gender                                                                           

In [87]:
df_merge.shape #Check how much column and rows

(10186, 24)

# Cleaning Merge Dataset

### Check Values and Typos

In [88]:
# Check Values and Typos
print(df_merge['Attrition_Flag'].unique())
print(df_merge['Gender'].unique())
print(df_merge['Education_Level'].unique())
print(df_merge['Marital_Status'].unique())
print(df_merge['Income_Category'].unique())
print(df_merge['Card_Category'].unique())

['Existing Customer' 'Attrited Customer']
['M' 'F' nan]
['High School' 'Graduate' 'Uneducated' 'Unknown' 'College' 'Post-Graduate'
 'Doctorate']
['Married' 'Single' 'Unknown' 'Divorced']
['$60K - $80K' '60ð\x9d\x90¾âˆ’ 80K' 'Less than $40K' '$80K - $120K'
 '$40K - $60K' '$120K +' 'Unknown']
['Blue' 'Gold' 'Silver' 'Platinum']


In [89]:
df_merge['Income_Category'].value_counts()

Less than $40K    3587
$40K - $60K       1795
$80K - $120K      1541
$60K - $80K       1406
Unknown           1115
$120K +            727
60𝐾− 80K       15
Name: Income_Category, dtype: int64

In [90]:
# replace typo on income category
df_merge['Income_Category'] = df_merge['Income_Category'].str.replace('60𝐾− 80K','$60K - $80K')

print(df_merge['Income_Category'].unique())

['$60K - $80K' 'Less than $40K' '$80K - $120K' '$40K - $60K' '$120K +'
 'Unknown']


### Treating Missing Values

In [91]:
df_merge.isnull().sum()

CLIENTNUM                                                                                                                              0
Attrition_Flag                                                                                                                         0
Customer_Age                                                                                                                          24
Gender                                                                                                                                 9
Dependent_count                                                                                                                        0
Education_Level                                                                                                                        0
Marital_Status                                                                                                                         0
Income_Category                          

In [92]:
# replace missing value on gender with unknown
df_merge['Gender'] = df_merge['Gender'].fillna('Unknown')

print(df_merge['Gender'].unique())


['M' 'F' 'Unknown']


In [93]:
# fillna age using mode
df_merge['Customer_Age'] = df_merge['Customer_Age'].fillna(df_merge['Customer_Age'].mode()[0])

In [94]:
#recheck null values
df_merge.isnull().sum()

CLIENTNUM                                                                                                                             0
Attrition_Flag                                                                                                                        0
Customer_Age                                                                                                                          0
Gender                                                                                                                                0
Dependent_count                                                                                                                       0
Education_Level                                                                                                                       0
Marital_Status                                                                                                                        0
Income_Category                                 

### Check Duplicate

In [95]:
#Check the duplicate
df_merge[df_merge.duplicated()]

CLIENTNUM     Attrition_Flag  Customer_Age   Gender  Dependent_count  \
21     714885258  Existing Customer         44.00        M                4   
23     709967358  Existing Customer         48.00        M                4   
25     753327333  Existing Customer         41.00        M                3   
27     806160108  Existing Customer         61.00        M                1   
29     709327383  Existing Customer         45.00        F                2   
31     806165208  Existing Customer         47.00        M                1   
33     708508758  Attrited Customer         62.00        F                0   
35     784725333  Existing Customer         41.00        M                3   
37     811604133  Existing Customer         47.00        F                4   
39     789124683  Existing Customer         54.00        M                2   
41     771071958  Existing Customer         41.00        F                3   
43     720466383  Existing Customer         59.00        M                1   
45     804424383  Existing Customer         63.00        M                1   
47     718813833  Existing Customer         44.00        F                3   
49     806624208  Existing Customer         47.00        M                4   
51     778348233  Existing Customer         53.00        M                3   
53     712991808  Existing Customer         53.00        M                2   
55     709029408  Existing Customer         41.00        M                4   
57     788658483  Existing Customer         53.00        F                2   
59     787937058  Existing Customer         58.00        M                0   
61     715318008  Existing Customer         55.00        F                1   
63     713962233  Existing Customer         55.00        F                3   
65     785432733  Existing Customer         42.00        F                4   
67     715190283  Existing Customer         57.00        F                1   
69     708300483  Attrited Customer         66.00        F                0   
10163  638805381  Existing Customer         44.00        M                3   
10165  518804382  Existing Customer         44.00        F                3   
10167  619912858  Existing Customer         44.00        M                4   
10169  539136358  Existing Customer         44.00        F                3   
10171  446911858  Existing Customer         44.00        M                4   
10173  479411858  Existing Customer         44.00  Unknown                4   
10175  539506333  Existing Customer         44.00  Unknown                4   
10177  533106332  Existing Customer         44.00  Unknown                4   

      Education_Level Marital_Status Income_Category Card_Category  \
21            Unknown        Unknown    $80K - $120K          Blue   
23      Post-Graduate         Single    $80K - $120K          Blue   
25            Unknown        Married    $80K - $120K          Blue   
27        High School        Married     $40K - $60K          Blue   
29           Graduate        Married         Unknown          Blue   
31          Doctorate       Divorced     $60K - $80K          Blue   
33           Graduate        Married  Less than $40K          Blue   
35        High School        Married     $40K - $60K          Blue   
37            Unknown         Single  Less than $40K          Blue   
39            Unknown        Married    $80K - $120K          Blue   
41           Graduate         Single  Less than $40K          Blue   
43        High School        Unknown     $40K - $60K          Blue   
45            Unknown        Married     $60K - $80K          Blue   
47         Uneducated         Single         Unknown          Blue   
49        High School        Married     $40K - $60K          Blue   
51            Unknown        Married    $80K - $120K          Blue   
53         Uneducated        Married     $60K - $80K          Blue   
55           Graduate        Married     $60K - $80K        

In [96]:
# delete duplicate
df_cleaned = df_merge.copy()
df_cleaned.drop_duplicates(subset ='CLIENTNUM', keep = 'first')

CLIENTNUM     Attrition_Flag  Customer_Age   Gender  Dependent_count  \
0      768805383  Existing Customer         45.00        M                3   
3      818770008  Existing Customer         49.00        F                5   
4      713982108  Existing Customer         51.00        M                3   
5      769911858  Existing Customer         40.00        F                4   
8      709106358  Existing Customer         40.00        M                3   
...          ...                ...           ...      ...              ...   
10181  532236358  Existing Customer         44.00        M                3   
10182  442211858  Existing Customer         44.00        M                4   
10183  479412258  Existing Customer         44.00  Unknown                4   
10184  539226333  Existing Customer         44.00  Unknown                4   
10185  522106332  Existing Customer         44.00  Unknown                4   

      Education_Level Marital_Status Income_Category Card_Category  \
0         High School        Married     $60K - $80K          Blue   
3            Graduate         Single  Less than $40K          Blue   
4            Graduate        Married    $80K - $120K          Blue   
5         High School        Unknown  Less than $40K          Blue   
8          Uneducated        Married     $60K - $80K          Blue   
...               ...            ...             ...           ...   
10181      Uneducated        Married     $60K - $80K          Gold   
10182     High School        Unknown  Less than $40K          Blue   
10183     High School        Unknown  Less than $40K          Blue   
10184     High School        Unknown  Less than $40K          Blue   
10185     High School        Unknown  Less than $40K          Blue   

       Months_on_book  ...  Credit_Limit_y  Total_Revolving_Bal  \
0                  39  ...        12691.00                  777   
3                  44  ...         8256.00                  864   
4                  36  ...         3418.00                    0   
5                  34  ...         3313.00                 2517   
8                  21  ...         4716.00                    0   
...               ...  ...             ...                  ...   
10181              21  ...          116.00                    0   
10182              34  ...          333.00                 2517   
10183              34  ...          333.00                 2517   
10184              34  ...          333.00                 2517   
10185              34  ...          333.00                 2517   

       Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
0             11914.00                  1.33             1144              42   
3              7392.00                  1.54             1291              33   
4              3418.00                  2.59             1887              20   
5               796.00                  1.41             1171              20   
8              4716.00                  2.17              816              28   
...                ...                   ...              ...             ...   
10181          4716.00                  2.17              816              28   
10182           796.00                  1.41             1171              20   
10183           796.00                  1.41             1171              20   
10184           796.00                  1.41             1171              20   
10185           796.00                  1.41             1171              20   

       Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  \
0                     1.62                   0.06   
3                     3.71                   0.10   
4                     2.33                   0.00   
5                     2.33                   0.76   
8                     2.50                   0.00   
...                    ...                    ...   
10181                 2.50                   0.00   
10182          

After delete duplicate total rows is 10148 (before 10181 rows)

# EDA

In [97]:
#Preparing a copy of df_merge for EDA
df_eda = df_cleaned.copy()
df_eda.head()

# Rounds a decimal number to 3 digits after a comma
pd.set_option('display.float_format', lambda x: '%.2f' % x)

numeric = ['Dependent_count', 'Credit_Limit_x', 'Credit_Limit_y','Total_Revolving_Bal', 'Avg_Open_To_Buy',
            'Total_Amt_Chng_Q4_Q1', 'Total_Trans_Amt', 'Total_Trans_Ct','Total_Ct_Chng_Q4_Q1','Avg_Utilization_Ratio',
            'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1',
            'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2']

desc_eda = df_eda[numeric].describe()
desc_eda.loc['kurtosis'] = df_eda[numeric].kurt() #show kurtosis statistic
desc_eda.loc['skewness'] = df_eda[numeric].skew() #show Skewness statistic
desc_eda.loc['variance'] = df_eda[numeric].var() #show variance statistic
desc_eda.round(2)
desc_eda

Dependent_count  Credit_Limit_x  Credit_Limit_y  \
count            10186.00        10186.00        10186.00   
mean                 2.35         8617.51         8617.51   
std                  1.30         9077.89         9077.89   
min                  0.00          116.00          116.00   
25%                  1.00         2548.25         2548.25   
50%                  2.00         4546.00         4546.00   
75%                  3.00        11089.75        11089.75   
max                  5.00        34516.00        34516.00   
kurtosis            -0.69            1.82            1.82   
skewness            -0.03            1.67            1.67   
variance             1.69     82408139.58     82408139.58   

          Total_Revolving_Bal  Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  \
count                10186.00         10186.00              10186.00   
mean                  1164.74          7459.56                  0.76   
std                    816.13          9076.46                  0.22   
min                      0.00             3.00                  0.00   
25%                    395.25          1320.00                  0.63   
50%                   1278.00          3473.00                  0.74   
75%                   1785.75          9882.25                  0.86   
max                   2517.00         34516.00                  3.40   
kurtosis                -1.15             1.81                  9.72   
skewness                -0.15             1.66                  1.79   
variance            666062.36      82382164.26                  0.05   

          Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
count            10186.00        10186.00             10186.00   
mean              4385.35           64.64                 0.72   
std               3396.21           23.59                 0.25   
min                510.00           10.00                 0.00   
25%               2137.25           45.00                 0.58   
50%               3889.00           67.00                 0.70   
75%               4738.00           81.00                 0.82   
max              18484.00          139.00                 3.71   
kurtosis             3.91           -0.38                15.92   
skewness             2.04            0.15                 2.39   
variance      11534212.69          556.52                 0.06   

          Avg_Utilization_Ratio  \
count                  10186.00   
mean                       0.28   
std                        0.28   
min                        0.00   
25%                        0.02   
50%                        0.18   
75%                        0.50   
max                        1.00   
kurtosis                  -0.81   
skewness                   0.72   
variance                   0.08   

          Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
count                                              10186.00                                                                                    
mean                                                   0.16                                                                                    
std                                                    0.36                                                                                    
min                                                    0.00                                                                                    
25%                                                    0.00                                                                                    
50%                                                    0.00                                                                                    
75%                                                    0.00                                                                                    
max                                           

In [98]:
df_eda[['Attrition_Flag','Gender','Education_Level','Marital_Status','Income_Category','Card_Category']].describe()

Attrition_Flag Gender Education_Level Marital_Status  \
count               10186  10186           10186          10186   
unique                  2      3               7              4   
top     Existing Customer      F        Graduate        Married   
freq                 8557   5380            3140           4713   

       Income_Category Card_Category  
count            10186         10186  
unique               6             4  
top     Less than $40K          Blue  
freq              3587          9489

### Average of customer's age with income 80k-120k

In [99]:
# Filter customer age based on income category
filtered_df_eda = df_eda[df_eda['Income_Category'].isin(['$80K - $120K'])]

# Calculate average age
average_age = filtered_df_eda['Customer_Age'].mean()

# Print the average age
print("Average of customer's age with income $80K - $120K:", average_age)


Average of customer's age with income $80K - $120K: 46.440622972096044


### Number of customers based on gender and education level

In [100]:
#Number of Customers based on Gender
user_gender = df_eda.groupby('Gender')['CLIENTNUM'].count()
user_gender = user_gender.reset_index()
user_gender

Gender  CLIENTNUM
0        F       5380
1        M       4797
2  Unknown          9

In [101]:
#Number of Customers based on Education Level
user_education_level = df_eda.groupby('Education_Level')['CLIENTNUM'].count()
user_education_level = user_education_level.reset_index()
user_education_level

Education_Level  CLIENTNUM
0         College       1015
1       Doctorate        453
2        Graduate       3140
3     High School       2042
4   Post-Graduate        517
5      Uneducated       1494
6         Unknown       1525

# Create New Column

### Added Average Transaction Amount column

In [102]:
df_eda['Avg_Trans_Amt'] = df_eda['Total_Trans_Amt'] / df_eda['Total_Trans_Ct']

In [103]:
#Check new column
df_eda.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10186 entries, 0 to 10185
Data columns (total 25 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10186 non-null  int64  
 1   Attrition_Flag                                                                                                                      10186 non-null  object 
 2   Customer_Age                                                                                                                        10186 non-null  float64
 3   Gender                                                                           